pkg/apiserver/apiserver.go

```diff
@@ -36,49 +38,133 @@ import (
+	"github.com/pingcap-incubator/tidb-dashboard/pkg/pd"
 	"github.com/pingcap-incubator/tidb-dashboard/pkg/tidb"
+	utils2 "github.com/pingcap-incubator/tidb-dashboard/pkg/utils"
+)
+
+var (
+	once sync.Once
 )
 
-var once sync.Once
+type PDDataProviderConstructor func(*config.Config, *http.Client, *clientv3.Client) *keyvisualregion.PDDataProvider
 
-type Services struct {
-	Store         *dbstore.DB
-	TiDBForwarder *tidb.Forwarder
-	KeyVisual     *keyvisual.Service
-	EtcdProvider  pd.EtcdProvider
-	HTTPClient    *http.Client
+type Service struct {
+	app    *fx.App
+	status *utils2.ServiceStatus
+
+	config            *config.Config
+	newPDDataProvider PDDataProviderConstructor
+	stoppedHandler    http.Handler
+
+	apiHandlerEngine *gin.Engine
 }
 
-func Handler(apiPrefix string, config *config.Config, services *Services) http.Handler {
+func newAPIHandlerEngine() (apiHandlerEngine *gin.Engine, endpoint *gin.RouterGroup, newTemplate utils2.NewTemplateFunc) {
+	apiHandlerEngine = gin.New()
+	apiHandlerEngine.Use(cors.AllowAll())
+	apiHandlerEngine.Use(gzip.Gzip(gzip.BestSpeed))
+	apiHandlerEngine.Use(utils.MWHandleErrors())
+
+	endpoint = apiHandlerEngine.Group("/dashboard/api")
+
+	newTemplate = func(name string) *template.Template {
+		return template.New(name).Funcs(apiHandlerEngine.FuncMap)
+	}
+
+	return
+}
+
+func NewService(cfg *config.Config, stoppedHandler http.Handler, newPDDataProvider PDDataProviderConstructor) *Service {
 	once.Do(func() {
 		// These global modification will be effective only for the first invoke.
+		_ = godotenv.Load()
 		gin.SetMode(gin.ReleaseMode)
 	})
 
-	r := gin.New()
-	r.Use(cors.AllowAll())
-	r.Use(gin.Recovery())
-	r.Use(gzip.Gzip(gzip.BestSpeed))
-	r.Use(utils.MWHandleErrors())
+	return &Service{
+		status:            utils2.NewServiceStatus(),
+		config:            cfg,
+		newPDDataProvider: newPDDataProvider,
+		stoppedHandler:    stoppedHandler,
+	}
+}
+
+func Register(r *gin.RouterGroup, s *Service) {
+	endpoint := r.Group("/dashboard/api")
+	endpoint.Use(s.status.MWHandleStopped(gin.WrapH(s.stoppedHandler)))
+	endpoint.Any("/*any", s.handler)
+}
+
+func (s *Service) Start(ctx context.Context) error {
+	s.app = fx.New(
+		fx.Logger(utils2.NewFxPrinter()),
+		fx.Provide(
+			newAPIHandlerEngine,
+			s.provide,
+			s.newPDDataProvider,
+			dbstore.NewDBStore,
+			pd.NewEtcdClient,
+			tidb.NewForwarderConfig,
+			tidb.NewForwarder,
+			http2.NewHTTPClientWithConf,
+			user.NewAuthService,
+			foo.NewService,
+			info.NewService,
+			clusterinfo.NewService,
+			profiling.NewService,
+			logsearch.NewService,
+			statement.NewService,
+			diagnose.NewService,
+			keyvisual.NewService,
+		),
+		fx.Populate(&s.apiHandlerEngine),
+		fx.Invoke(
+			user.Register,
+			foo.Register,
+			info.Register,
+			clusterinfo.Register,
+			profiling.Register,
+			logsearch.Register,
+			statement.Register,
+			diagnose.Register,
+			keyvisual.Register,
+			// Must be at the end
+			s.status.Invoke,
```

Comment:
prefer to call it `s.runningStatus.HookLifecycle`

Reply:
I think it is more unified to change to `Register`. It may depend on more than just `Lifecycle`